### PR TITLE
perf: add searchHint to getSeriesIndex

### PR DESCRIPTION
**Pull request by liaol**
The `phlaredb.getSeriesIndex()` function is a performance bottleneck during "flushing", consuming approximately 12% of CPU.  

https://github.com/grafana/pyroscope/blob/4a2d0a8ed450c9b1e2bd249347b197efd723e99b/pkg/phlaredb/profiles.go#L43-L54
Since `rowRangesWithSeriesIndex` is strictly ordered and `rowNum` increments sequentially, implementing a `searchHint` is beneficial. This optimization reduced CPU usage of `getSeriesIndex` to 5% and reduced the "flushing" operation's duration from minutes to seconds in our tests.

Furthermore, if `seriesIDRowsRewriter.ReadRows()` has no data-race, adding `searchHint` to `seriesIDRowsRewriter`  struct would further improve efficiency.
https://github.com/grafana/pyroscope/blob/4a2d0a8ed450c9b1e2bd249347b197efd723e99b/pkg/phlaredb/profile_store.go#L538-L553

### CODE REVIEW

**Comment by korniltsev** [comment]
Can you share a bit more about the size of your blocks? How many rowgroups, rows, series does it have? Something tells me you have too many series.

> Furthermore, if seriesIDRowsRewriter.ReadRows() has no data-race, adding searchHint to seriesIDRowsRewriter struct would further improve efficiency.

I don't see no concurrency in there during flushing (Except `(r *seriesIDRowsRewriter) SeekToRow` which is not used and imo should be removed). Let's implement this idea, should be way better as the buffer for the reader is only 64 .

**Comment by liaol** [comment]
@korniltsev Yes, we have big series, the  size of `rowRangeWithSeriesIndex` is over 200K. So the seachHint will save over 63 * 200K operations.
I will move  `searchHint` to `seriesIDRowsRewriter struct` later.

File: pkg/phlaredb/profiles.go [hunk 1]
```diff
@@ -40,13 +40,26 @@ type rowRangeWithSeriesIndex struct {
...
 		// it is possible that the series is not existing
 		if rg.rowRange == nil {
 			continue
 		}
 		if rg.rowNum <= rowNum && rg.rowNum+int64(rg.length) > rowNum {
+			*searchHint = i
+			return rg.seriesIndex
+		}
+	}
+	for i := 0; i < *searchHint; i++ {
```
Comment (korniltsev) [inline pkg/phlaredb/profiles.go]: Is the second loop needed?

Comment (liaol) [reply pkg/phlaredb/profiles.go]: Yes, for the corner case which not increments sequentially.

**Comment by liaol** [comment]
@korniltsev I've moved `searchHint` to `seriesIDRowsRewriter`. Please review again.

File: pkg/phlaredb/profiles.go (hunk 1 above)
Comment (korniltsev) [reply pkg/phlaredb/profiles.go]: Do you know how to reproduce this case? It's worth adding a comment what the corner case is and how to trigger it. Or maybe even a unit test so that we do not remove this in the future accidentally.

Comment (liaol) [reply pkg/phlaredb/profiles.go]: For example, consider row numbers 1, 2, 3, 4, 3, which may not actually exist. The second loop can maintain the consistency of `getSeriesIndex` before and after modification.

Comment (korniltsev) [reply pkg/phlaredb/profiles.go]: I don't understand. What is 1,2,3,4,3 and what does not exist. Can you add a test? Or at least source code comment explaining why the second loop needed.

**Review by korniltsev — CHANGES_REQUESTED** [review]
While the second loop will not hurt, I'd like to have understanding myself and documented in the form of a comment or a test in the source code

**Comment by liaol** [comment]
@korniltsev Sorry for not explaining clearly, I have submitted the unit tests and comments.

File: pkg/phlaredb/profile_test.go [hunk 2]
```diff
@@ -279,3 +279,93 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+		{rowRange: &rowRange{rowNum: 0, length: 5}, seriesIndex: 1},
+		{rowRange: &rowRange{rowNum: 5, length: 5}, seriesIndex: 2},
+	}
+	searchHint := 0
+
+	idx := ranges.getSeriesIndex(6, &searchHint)
+	assert.Equal(t, uint32(2), idx)
+	assert.Equal(t, 1, searchHint)
+
+	// rowNum is not increasing, so it will not hit the first loop, and will hit the second loop
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: I don't get it. why would it decrease? Looking at the code it goes from zero to number of rows in rowgroup

https://github.com/grafana/pyroscope/blob/5b90569d50e869fce59e0784cee1044737e5ec06/pkg/phlaredb/profile_store.go#L546-L552

File: pkg/phlaredb/profile_test.go [hunk 3]
```diff
@@ -279,3 +279,93 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+			searchHint: 1, // will skip 0 in main loop, hit in second loop
+			expectIdx:  0,
+		},
+		{
+			name: "searchHint out of range",
+			ranges: rowRangesWithSeriesIndex{
+				{rowRange: &rowRange{rowNum: 0, length: 5}, seriesIndex: 1},
+			},
+			rowNum:     2,
+			searchHint: 100, // should reset to 0
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: I don't get it. Why would searchHint be 100 if the rowgroup only have a single series?

**Comment by liaol** [comment]
You're right, the second loop is unnecessary because rowNum is always incrementing.
I have removed the second loop and the unnecessary unit tests.

File: pkg/phlaredb/profile_test.go [hunk 4]
```diff
@@ -279,3 +279,67 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+			ranges: rowRangesWithSeriesIndex{
+				{rowRange: &rowRange{rowNum: 0, length: 5}, seriesIndex: 1},
+				{rowRange: &rowRange{rowNum: 5, length: 5}, seriesIndex: 2},
+			},
+			rowNum:     6,
+			searchHint: 1,
+			expectIdx:  1,
+		},
+		{
+			name: "second loop hit",
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: there is no second loop

File: pkg/phlaredb/profile_test.go [hunk 5]
```diff
@@ -279,3 +279,67 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+	}{
+		{
+			name: "main loop hit",
+			ranges: rowRangesWithSeriesIndex{
+				{rowRange: &rowRange{rowNum: 0, length: 5}, seriesIndex: 1},
+				{rowRange: &rowRange{rowNum: 5, length: 5}, seriesIndex: 2},
+			},
+			rowNum:     6,
+			searchHint: 1,
+			expectIdx:  1,
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: there is something wrong with the testate. rownum should point to row range with index 1 with seriesIndex 2

File: pkg/phlaredb/profile_test.go [hunk 6]
```diff
@@ -279,3 +279,67 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
 
 }
+
+func Test_rowRangesWithSeriesIndex_getSeriesIndex(t *testing.T) {
+	testCases := []struct {
+		name        string
+		ranges      rowRangesWithSeriesIndex
+		rowNum      int64
+		searchHint  int
+		expectIdx   int
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: we should rename this to  expectedSeriesIndex, to not confuse with any other indexes

File: pkg/phlaredb/profile_test.go [hunk 7]
```diff
@@ -279,3 +279,67 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+		},
+		{
+			name: "second loop hit",
+			ranges: rowRangesWithSeriesIndex{
+				{rowRange: &rowRange{rowNum: 0, length: 5}, seriesIndex: 1},
+				{rowRange: &rowRange{rowNum: 5, length: 5}, seriesIndex: 2},
+			},
+			rowNum:     1,
+			searchHint: 1, // will skip 0 in main loop, hit in second loop
+			expectIdx:  0,
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: same here, rownum is in row range number 0 with seriesIndex 1. we should expect seriesIndex = 1 here, not 0

File: pkg/phlaredb/profile_test.go (hunk 5 above)
Comment (korniltsev) [reply pkg/phlaredb/profile_test.go]: we should expect seriesIndex2 here not 1

File: pkg/phlaredb/profile_test.go [hunk 8]
```diff
@@ -279,3 +279,67 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+		},
+		{
+			name: "nil rowRange skipped",
+			ranges: rowRangesWithSeriesIndex{
+				{rowRange: nil, seriesIndex: 1},
+				{rowRange: &rowRange{rowNum: 10, length: 5}, seriesIndex: 2},
+			},
+			rowNum:     12,
+			searchHint: 0,
+			expectIdx:  1,
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: similar to above, the expected value should be 2

File: pkg/phlaredb/profile_test.go [hunk 9]
```diff
@@ -279,3 +279,67 @@ func TestProfileIndex_Add_OutOfOrder(t *testing.T) {
...
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			searchHint := tc.searchHint
+			if tc.expectPanic {
+				assert.Panics(t, func() {
+					_ = tc.ranges.getSeriesIndex(tc.rowNum, &searchHint)
+				})
+			} else {
+				idx := tc.ranges.getSeriesIndex(tc.rowNum, &searchHint)
+				assert.Equal(t, tc.ranges[tc.expectIdx].seriesIndex, idx)
```
Comment (korniltsev) [inline pkg/phlaredb/profile_test.go]: I think this assertion is incorrect. `getSeriesIndex` returns `seriesIndex`, not index into the ranges slice.  This should be fixed.

**Review by korniltsev — CHANGES_REQUESTED** [review]

File: pkg/phlaredb/profile_test.go (hunk 9 above)
Comment (liaol) [reply pkg/phlaredb/profile_test.go]: all fixed.

**Review by korniltsev — APPROVED** [review]
